opam: Add new versions

ashermancinelli opened the pull request:
Also: set the build and install directories to the source directory because the build system unfortunately expects the `src_ext` directory to be under the current working directory when building the bundled third-party libraries, even when the configure script is run from another directory.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

ashermancinelli committed:
opam: Add new versions

Also: set the build and install directories to the source directory
because the build system unfortunately expects the `src_ext` directory
to be under the current working directory when building the bundled
third-party libraries, even when the configure script is run from
another directory.

ashermancinelli commented:
To verify the checksums I've given are correct:
```shell
$ rm ./opam*.tar.gz
$ for ver in 2.1.4 2.1.5 2.1.6 2.2.0 2.2.1; do
>     url="https://github.com/ocaml/opam/releases/download/${ver}/opam-full-${ver}.tar.gz"
>     echo $url
>     nohup wget "$url" 2>/dev/null
>     sha256sum opam-full-${ver}.tar.gz
> done
https://github.com/ocaml/opam/releases/download/2.1.4/opam-full-2.1.4.tar.gz
1643609f4eea758eb899dc8df57b88438e525d91592056f135e6e045d0d916cb  opam-full-2.1.4.tar.gz
https://github.com/ocaml/opam/releases/download/2.1.5/opam-full-2.1.5.tar.gz
09f8d9e410b2f5723c2bfedbf7970e3b305f5017895fcd91759f05e753ddcea5  opam-full-2.1.5.tar.gz
https://github.com/ocaml/opam/releases/download/2.1.6/opam-full-2.1.6.tar.gz
d2af5edc85f552e0cf5ec0ddcc949d94f2dc550dc5df595174a06a4eaf8af628  opam-full-2.1.6.tar.gz
https://github.com/ocaml/opam/releases/download/2.2.0/opam-full-2.2.0.tar.gz
39334f36adbe280683487cf204b7b0642080fc5965747f7d6f7cc7b83cd7a192  opam-full-2.2.0.tar.gz
https://github.com/ocaml/opam/releases/download/2.2.1/opam-full-2.2.1.tar.gz
07ad3887f61e0bc61a0923faae16fcc141285ece5b248a9e2cd4f902523cc121  opam-full-2.2.1.tar.gz
```

scemama commented on var/spack/repos/builtin/packages/opam/package.py:
```diff
@@ -36,10 +41,18 @@ class Opam(AutotoolsPackage):
     depends_on("ocaml", type="build", when="@2.0.0:")
 
-    parallel = False
+    # TODO: Is this really the correct constraint? I can't find any issues with
+    # building newer opams in parallel, and the constraint was added in the
+    # package's first commit with no message.
+    with when("@:2.0.0"):
+        parallel = False
```
`make` is a frontend to the `dune` builder, which is parallel. So there is no advantage to run `make` in parallel.
Can you please add this as a comment? Thanks ;-)

scemama replied:
I found this thread:
https://github.com/ocaml/opam/issues/3585
So in my opinion, it is prefereable to leave `parallel = False`

ashermancinelli replied:
Thank you for the context! I'll add a comment.

scemama replied:
Thanks, this comment can be marked as resolved

scemama reviewed (approved):
Thanks a lot.

ashermancinelli committed:
Add comment about not building in parallel

@scemama pointed out that 'make' just calls 'dune' which is already
parallel, so make itself should not have more than one job.

bernhardkaindl reviewed (dismissed):
Built opam@2.2.1 on Ubuntu 22.04 with the newly merged ocaml@5.2.0.

However, opam@2.1.4, @2.1.5 and @2.1.6 failed to build:
```ml
     107    src_ext/dune-local/dune.exe build --profile=release --root .  --promote-install-files -- opam-installer.install opam.install
     108    File "src/tools/dune", line 29, characters 27-35:
     109    29 |   (libraries   opam-format cmdliner)
     110                                    ^^^^^^^^
  >> 111    Error: Library "cmdliner" not found.
     112    -> required by _build/default/src/tools/opam_installer.exe
     113    -> required by _build/install/default/bin/opam-installer
     114    -> required by _build/default/opam-installer.install
     115    File "src/client/dune", line 6, characters 64-72:
     116    6 |   (libraries   opam-state opam-solver opam-repository re base64 cmdliner
     117                                                                        ^^^^^^^^
  >> 118    Error: Library "cmdliner" not found.
```

bernhardkaindl committed:
opam@:2.1 still need 'make lib-ext' for cmdliner

Signed-off-by: Bernhard Kaindl <bernhard.kaindl@cloud.com>

bernhardkaindl reviewed (approved):
Fixed 2.1.x: Versions up to 2.1.6 still need `make lib-ext` for cmdliner.
